src/uwtools/rocoto.py

```diff
@@ -181,14 +158,57 @@ def _add_task_dependency(self, e: Element, config: dict) -> None:
         :param e: The parent element to add the new element to.
         :param config: Configuration data for this element.
         """
+        operands, operators, strequality = self._dependency_constants
         e = SubElement(e, STR.dependency)
-        for key, block in config.items():
-            tag, _ = self._tag_name(key)
-            if tag == STR.taskdep:
-                self._set_attrs(SubElement(e, STR.taskdep), block)
+
+        for tag, block in config.items():
+            tag, _ = self._tag_name(tag)
+            if tag in operands:
+                self._add_task_dependency_operand(e, tag=tag, block=block)
+            elif tag in operators:
+                self._add_task_dependency_operator(e, config={tag: block})
+            elif tag in strequality:
+                self._add_task_dependency_strequality(e, tag=tag, block=block)
+            else:
+                raise UWConfigError("Unhandled dependency type %s" % tag)
+
+    def _add_task_dependency_operand(self, e: Element, tag: str, block: dict) -> None:
+        """
+        :param e: The parent element to add the new element to.
+        :param tag: Configuration new element to add.
+        :param block: Configuration data for the tag.
+        """
+        self._set_attrs(SubElement(e, tag), block)
+
+    def _add_task_dependency_operator(self, e: Element, config: dict) -> None:
+        """
+        Add an operator element to the <dependency>.
+
+        :param e: The parent element to add the new element to.
+        :param config: Configuration data for this element.
+        """
+
+        operands, operators, strequality = self._dependency_constants
+
+        for tag, block in config.items():
+            tag, _ = self._tag_name(tag)
+            if tag in operators:
+                self._add_task_dependency_operator(SubElement(e, tag), config=block)
+            elif tag in operands:
+                self._add_task_dependency_operand(e, tag=tag, block=block)
+            elif tag in strequality:
+                self._add_task_dependency_strequality(e, tag=tag, block=block)
             else:
                 raise UWConfigError("Unhandled dependency type %s" % tag)
 
+    def _add_task_dependency_strequality(self, e: Element, tag: str, block: dict) -> None:
```

Comment:
Can we sort all of the arguments alphabetically as well?

Reply:
I have to (partially) oppose that suggestion, with rationale: It's a common API design strategy to present arguments in a consistent way so that users have a good idea what to expect. For example, if you've got a collection of functions/methods that all operate on the same kind of thing, but maybe with various extra arguments specific to each's functionality, a consistent design would keep the common argument(s) first in the argument list, so that that would be an invariant across the whole collection. That's (mostly? completely?) the case here, where we've got a bunch of methods that all operate on `Element` objects. The `Element` argument should come first, because it's the invariant across all the methods. If there were multiple arguments that are used across all the calls, sorting those might make sense, though it might also make sense to order them according to some kind of abstract hierarchy -- e.g. if you had functions that all did things related to flowers and their colors, you might have `foo(flower, color, ..., ...)` with `flower` and `color` in the "wrong" order because "flower" is somehow a larger or more crucial concept the "color" for whatever domain we're working in.

I want to write something more, but am going to submit this first to make sure, hopefully, that we don't merge this PR with the `Element` argument `e` demoted...

Reply:
Another reason for ordering arguments in a specific way is to use what in functional programming is called "currying" or "partial application". Python supports this with the `partial()` function from the `functools` library. Currying allows you to turn a function with `n` arguments into a function with `n-1` arguments, where the first argument is fixed by the currying operation. For example, in Python, you might have
```
>>> def flower(name, color):
...     print(f"This {name} is {color}.")
... 
>>> flower("rose", "red")
This rose is red.
```
But if you wanted to do a lot of things with roses, you might
```
>>> from functools import partial
>>> rose = partial(flower, "rose")
>>> rose("white")
This rose is white.
>>> rose("yellow")
This rose is yellow.
```
This pattern is very common in functional programming, and quite useful in Python sometimes. Success, though, requires that arguments be ordered in such a way that you can "fix" the leading argument with `partial()` in a useful way, such that you usually want the argument(s) representing the "main concept(s)" to come first. You can, of course, curry multiple times to derive more and more specific functions.

So, fully ordering argument lists lexicographically may not make sense is some cases, and I think our case here is one of those. Past the initial `Element` argument I'm not sure I care, but if e.g. every method take `e`, and then _a lot_ of elements take a `config`, and just _some_ methods take further arguments, `config` might similarly be such a common concept that users will find it ergonomic to see `f(e, config, ... other args ...)`